### PR TITLE
Fix cached tracing for `IO.delay`

### PR DESCRIPTION
**Pull request by armanbilge**
The first commit demonstrates the issue described in #2225, note the test fails only for Scala 3.

### CODE REVIEW

File: core/shared/src/main/scala/cats/effect/IO.scala [hunk 1]
```diff
@@ -931,7 +931,7 @@ object IO extends IOCompanionPlatform with IOLowPriorityImplicits {
    */
   def apply[A](thunk: => A): IO[A] = {
     val fn = () => thunk
-    Delay(fn, Tracing.calculateTracingEvent(fn.getClass))
+    Delay(fn, Tracing.calculateTracingEvent(tracing.ThunkClassRetriever(thunk)))
```
Comment (vasilmkd) [inline core/shared/src/main/scala/cats/effect/IO.scala]: Inline the lambda in `Delay` and remove the braces around the method.

File: core/jvm/src/main/scala/cats/effect/tracing/ThunkClassRetriever.scala [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+private abstract class ThunkClassRetriever {
+  def getClassOf[A](thunk: => A): Class[_]
+}
+
+private[effect] object ThunkClassRetriever {
+  private val impl: ThunkClassRetriever = new ThunkClassRetrieverImpl
+
+  def apply[A](thunk: => A): Class[_] = impl.getClassOf(thunk)
+}
```
Comment (vasilmkd) [inline core/jvm/src/main/scala/cats/effect/tracing/ThunkClassRetriever.scala]: Can you move this into `Tracing.scala`?

Comment (vasilmkd) [reply core/jvm/src/main/scala/cats/effect/tracing/ThunkClassRetriever.scala]: And rename the `apply` method to getThunkClass or something.

Comment (armanbilge) [reply core/jvm/src/main/scala/cats/effect/tracing/ThunkClassRetriever.scala]: Err, wait, not sure. `Tracing.scala` is cross-platform but we can't implement this on JS.

Comment (vasilmkd) [reply core/jvm/src/main/scala/cats/effect/tracing/ThunkClassRetriever.scala:27]: `Platform`.

File: core/js/src/main/scala/cats/effect/tracing/ThunkClassRetriever.scala [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cats.effect.tracing
+
+private[tracing] class ThunkClassRetrieverImpl extends ThunkClassRetriever {
```
Comment (vasilmkd) [inline core/js/src/main/scala/cats/effect/tracing/ThunkClassRetriever.scala]: No need for `private[tracing]`, just `private`.

Comment (armanbilge) [reply core/js/src/main/scala/cats/effect/tracing/ThunkClassRetriever.scala]: If it's just `private`, then we can't do `new` in `Tracing` I think.

Comment (vasilmkd) [reply core/js/src/main/scala/cats/effect/tracing/ThunkClassRetriever.scala]: Also please make the filename `ThunkClassRetrieverImpl`.

Comment (armanbilge) [reply core/js/src/main/scala/cats/effect/tracing/ThunkClassRetriever.scala]: Oh I guess we can :) cool

**Review by djspiewak — COMMENTED** [review]
So this makes a lot of sense, and is also disheartening because it means that Scala 3 isn't optimizing thunk-to-function conversion, and thus Delay nodes are more expensive on that platform. I actually wonder if it would be even better to generalize this machinery slightly, and rather than returning the Class, return a Function0. We can still get the class from it, but effectively we would be cheating the compiler into implementing the optimization it previously did naturally.

**Comment by armanbilge** [comment]
> I actually wonder if it would be even better to generalize this machinery slightly, and rather than returning the Class, return a Function0

Clever! Why not?

**Comment by vasilmkd** [comment]
Please make it a part of tracing...

**Comment by armanbilge** [comment]
Lol!

**Comment by vasilmkd** [comment]
Or keep it in `cats.effect`. Just not in a subpackage.

File: core/jvm/src/main/java/cats/effect/util/ThunkImpl.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+ * limitations under the License.
+ */
+
+package cats.effect.util;
+
+class ThunkImpl extends Thunk {
+  public <A> scala.Function0<A> asFunction0(scala.Function0<A> thunk) {
+    return thunk;
+  }
+}
```
Comment (vasilmkd) [inline core/jvm/src/main/java/cats/effect/util/ThunkImpl.java]: Let's add a comment/javadoc on this class explaining the need for its existence. Mention the differences between Scala 2 and Scala 3.

File: core/jvm/src/main/java/cats/effect/ThunkImpl.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+package cats.effect;
+
+/**
+ * A utility to convert a by-name `thunk: => A` to a `Function0[A]` (its binary representation).
+ * Scala 2 performs this optimization automatically but on Scala 3 the thunk is wrapped inside of a new `Function0`.
+ * See https://github.com/typelevel/cats-effect/pull/2226
+ */
+class ThunkImpl extends Thunk {
+  public <A> scala.Function0<A> asFunction0(scala.Function0<A> thunk) {
```
Comment (vasilmkd) [inline core/jvm/src/main/java/cats/effect/ThunkImpl.java]: Add an `@Override` here.

File: core/shared/src/main/scala/cats/effect/tracing/ThunkClassRetriever.scala [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cats.effect.tracing
+
+private abstract class ThunkClassRetriever {
+  def getClassOf[A](thunk: => A): Class[_]
+}
```
Comment (vasilmkd) [inline core/shared/src/main/scala/cats/effect/tracing/ThunkClassRetriever.scala]: This is a leftover from the previous encoding. I think this is the final change left to do.

**Review by vasilmkd — CHANGES_REQUESTED** [review]

**Review by vasilmkd — COMMENTED** [review]
Can we run a quick `DeepBindBenchmark.delay` _without_ tracing on Scala 2.13 and Scala 3? Before and after.

**Comment by armanbilge** [comment]
Command: `benchmarks/Jmh/run -wi 10 -i 10 -f 2 -t 1 DeepBindBenchmark.delay --jvmArgs -Dcats.effect.tracing.mode=none`

2.13 before:
```
[info] Benchmark                (size)   Mode  Cnt     Score    Error  Units
[info] DeepBindBenchmark.delay   10000  thrpt   20  3588.051 ± 75.498  ops/s
```

3 before:
```
[info] Benchmark                (size)   Mode  Cnt     Score    Error  Units
[info] DeepBindBenchmark.delay   10000  thrpt   20  2841.203 ± 45.684  ops/s
```

2.13 after:
```
[info] Benchmark                (size)   Mode  Cnt     Score    Error  Units
[info] DeepBindBenchmark.delay   10000  thrpt   20  3487.627 ± 87.979  ops/s
```

3 after:
```
[info] Benchmark                (size)   Mode  Cnt     Score    Error  Units
[info] DeepBindBenchmark.delay   10000  thrpt   20  3897.315 ± 39.551  ops/s
```

**Comment by armanbilge** [comment]
2.13 @ 7732752:
```
[info] Benchmark                (size)   Mode  Cnt     Score     Error  Units
[info] DeepBindBenchmark.delay   10000  thrpt   20  3453.852 ± 110.719  ops/s
```

3 @ 7732752:
```
[info] Benchmark                (size)   Mode  Cnt     Score     Error  Units
[info] DeepBindBenchmark.delay   10000  thrpt   20  3431.805 ± 216.061  ops/s
```

**Comment by armanbilge** [comment]
3 @ 7732752 (take 2):
```
[info] Benchmark                (size)   Mode  Cnt     Score     Error  Units
[info] DeepBindBenchmark.delay   10000  thrpt   20  3786.509 ± 129.755  ops/s
```

**Comment by vasilmkd** [comment]
Here are my results:

```
Baseline 2.13
Benchmark                (size)   Mode  Cnt     Score    Error  Units
DeepBindBenchmark.delay   10000  thrpt   20  3028.428 ± 84.643  ops/s

Baseline 3.0
Benchmark                (size)   Mode  Cnt     Score     Error  Units
DeepBindBenchmark.delay   10000  thrpt   20  2848.246 ± 204.663  ops/s

Java code 2.13
Benchmark                (size)   Mode  Cnt     Score    Error  Units
DeepBindBenchmark.delay   10000  thrpt   20  3116.695 ± 84.533  ops/s

Java code 3.0
Benchmark                (size)   Mode  Cnt     Score     Error  Units
DeepBindBenchmark.delay   10000  thrpt   20  3706.874 ± 365.418  ops/s

Cross-platform 2.13
Benchmark                (size)   Mode  Cnt     Score     Error  Units
DeepBindBenchmark.delay   10000  thrpt   20  3328.456 ± 338.778  ops/s

Cross-platform 3.0
Benchmark                (size)   Mode  Cnt     Score     Error  Units
DeepBindBenchmark.delay   10000  thrpt   20  3414.873 ± 111.307  ops/s
```

File: core/shared/src/main/scala/cats/effect/Thunk.scala [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cats.effect
+
+private object Thunk {
+  private[this] val impl =
```
Comment (vasilmkd) [inline core/shared/src/main/scala/cats/effect/Thunk.scala:25]: Can you please restore the comment that you put on the Java code.

**Review by vasilmkd — CHANGES_REQUESTED** [review]

**Review by vasilmkd — APPROVED** [review]
Great work. I can't tell you how happy I am that we get to keep cached tracing for `apply/delay`.

**Pull request merged**